components/scream/src/dynamics/homme/physics_dynamics_remapper.hpp

```diff
@@ -589,396 +994,493 @@ setup_boundary_exchange () {
-  const auto& phys_dims = phys_layout.dims();
-  switch (get_layout_type(phys_layout.tags())) {
-    case LayoutType::Scalar2D:
+  switch (phys_layout(i)) {
+    case 0:
     {
-      auto phys = phys_field.template get_reshaped_view<Real*>();
-      if (itl>=0) {
-        auto dyn  = dyn_field.template get_reshaped_view<Real**[NP][NP]>();
-        Kokkos::parallel_for(RangePolicy(0,num_cols),
-                             KOKKOS_LAMBDA(const int icol) {
+      auto phys = Unmanaged<KokkosTypes<DefaultDevice>::view<ScalarT*>> (reinterpret_cast<ScalarT*>(phys_ptrs(i).get()),
+                                                                         phys_dims(i,0));
+
+      const auto tr = Kokkos::TeamThreadRange(team, num_cols);
+      if (itl_fwd(i)>=0) {
+        auto dyn = Unmanaged<KokkosTypes<DefaultDevice>::view<ScalarT**[NP][NP]>> (reinterpret_cast<ScalarT*>(dyn_ptrs(i).get()),
+                                                                                   dyn_dims(i,0),
+                                                                                   dyn_dims(i,1));
+
+        const auto f = [&] (const int icol) {
           const auto& elgp = Kokkos::subview(lid2elgp,p2d(icol),Kokkos::ALL());
-          dyn(elgp[0],itl,elgp[1],elgp[2]) = phys(icol);
-        });
+          dyn(elgp[0],itl_fwd(i),elgp[1],elgp[2]) = phys(icol);
+        };
+        Kokkos::parallel_for(tr, f);
       } else {
-        auto dyn = dyn_field.template get_reshaped_view<Real*[NP][NP]>();
-        Kokkos::parallel_for(RangePolicy(0,num_cols),
-                             KOKKOS_LAMBDA(const int icol) {
+        auto dyn = Unmanaged<KokkosTypes<DefaultDevice>::view<ScalarT*[NP][NP]>> (reinterpret_cast<ScalarT*>(dyn_ptrs(i).get()),
+                                                                                  dyn_dims(i,0));
+
+        const auto tr = Kokkos::TeamThreadRange(team, num_cols);
+        const auto f = [&] (const int icol) {
           const auto& elgp = Kokkos::subview(lid2elgp,p2d(icol),Kokkos::ALL());
           dyn(elgp[0],elgp[1],elgp[2]) = phys(icol);
-        });
+        };
+        Kokkos::parallel_for(tr, f);
       }
       break;
     }
-    case LayoutType::Vector2D:
+    case 1:
     {
-      auto phys = phys_field.template get_reshaped_view<Real**>();
-      if (itl>=0) {
-        auto dyn = dyn_field.template get_reshaped_view<Real***[NP][NP]>();
-        const int dim = phys_dims[1];
-        Kokkos::parallel_for(RangePolicy(0,num_cols*dim),
-                             KOKKOS_LAMBDA(const int idx) {
-          const int icol = idx / dim;
-          const int idim = idx % dim;
+      auto phys = Unmanaged<KokkosTypes<DefaultDevice>::view<ScalarT**>> (reinterpret_cast<ScalarT*>(phys_ptrs(i).get()),
+                                                                          phys_dims(i,0),
+                                                                          phys_dims(i,1));
+
+      const int dim = phys_dims(i, 1);
```

Comment:
ABOUT PHYS_DIMS:

If phys dims were a 1d view of Kokkos::Array, we could do
```
// Outside the switch:
const auto& dims = phys_dims(i);
...
// inside the loops
const icol = idx / dims[0];
...
const icol = idx / (dims[0]*dims[1]);
```
without having dim1,dim2,.. I _think_ this _might_ reduce register pressure (and maybe allow an extra block to fit on the SM).

Reply:
An alternative to `Kokkos::Array` would be a POD type:
```
struct PhysDims {
  int dim0;
  int dim1;
  int dim2;
  int dim3;
};
```
And you could even throw in `int size;` as a member, so to get rid of `phys_dims_size`. I might actually prefer this to `Kokkos::Array`.

Reply:
I like the struct option best, I think. Let me see how clean it is.

Reply:
It's definitely more verbose.

Reply:
This issue with the struct is that at one place I need to access the last dimension, which would require a large if-block twice for both phys and dyn. What about a combo of the 2?
```
  struct Dims {
    int size;
    Kokkos::Array<int,6> dims;
  };
```

Reply:
Ah, cause you the idx to access is taken from a variable. I see. Then I guess what you suggest is a good solution.